In [1]:
# Data processing  
# -----------------------------------------------------------------------  
import pandas as pd  
import numpy as np

# Pandas options  
# -----------------------------------------------------------------------  
pd.options.display.max_colwidth = None
 
# Path configuration for custom module imports  
# -----------------------------------------------------------------------  
import sys  
sys.path.append('../')  # Adds the parent directory to the path for custom module imports  

# Ignore warnings  
# -----------------------------------------------------------------------  
import warnings  
warnings.filterwarnings("ignore")

# Custom functions
# -----------------------------------------------------------------------
from src.support_eda import *

In [8]:
df = pd.read_csv("../data/raw/Global_Superstore.csv", encoding='latin-1')

En primer lugar, renombramos las columnas para eliminar los espacios

In [10]:
df.columns = [col.replace(' ', '') for col in df.columns]

df.columns

Index(['RowID', 'OrderID', 'OrderDate', 'ShipDate', 'ShipMode', 'CustomerID',
       'CustomerName', 'Segment', 'City', 'State', 'Country', 'PostalCode',
       'Market', 'Region', 'ProductID', 'Category', 'Sub-Category',
       'ProductName', 'Sales', 'Quantity', 'Discount', 'Profit',
       'ShippingCost', 'OrderPriority'],
      dtype='object')

In [11]:
df.head()

RowID          OrderID   OrderDate    ShipDate      ShipMode CustomerID  \
0  32298   CA-2012-124891  31-07-2012  31-07-2012      Same Day   RH-19495   
1  26341    IN-2013-77878  05-02-2013  07-02-2013  Second Class   JR-16210   
2  25330    IN-2013-71249  17-10-2013  18-10-2013   First Class   CR-12730   
3  13524  ES-2013-1579342  28-01-2013  30-01-2013   First Class   KM-16375   
4  47221     SG-2013-4320  05-11-2013  06-11-2013      Same Day    RH-9495   

       CustomerName      Segment           City            State  ...  \
0       Rick Hansen     Consumer  New York City         New York  ...   
1     Justin Ritter    Corporate     Wollongong  New South Wales  ...   
2      Craig Reiter     Consumer       Brisbane       Queensland  ...   
3  Katherine Murray  Home Office         Berlin           Berlin  ...   
4       Rick Hansen     Consumer          Dakar            Dakar  ...   

          ProductID    Category Sub-Category  \
0   TEC-AC-10003033  Technology  Accessories   
1   FUR-CH-10003950   Furniture       Chairs   
2   TEC-PH-10004664  Technology       Phones   
3   TEC-PH-10004583  Technology       Phones   
4  TEC-SHA-10000501  Technology      Copiers   

                                                          ProductName  \
0  Plantronics CS510 - Over-the-Head monaural Wireless Headset System   
1                           Novimex Executive Leather Armchair, Black   
2                                   Nokia Smart Phone, with Caller ID   
3                                      Motorola Smart Phone, Cordless   
4                                      Sharp Wireless Fax, High-Speed   

      Sales Quantity Discount    Profit  ShippingCost  OrderPriority  
0  2309.650        7      0.0  762.1845        933.57       Critical  
1  3709.395        9      0.1 -288.7650        923.63       Critical  
2  5175.171        9      0.1  919.9710        915.49         Medium  
3  2892.510        5      0.1  -96.5400        910.16         Medium  
4  2832.960        8      0.0  311.5200        903.04       Critical  

[5 rows x 24 columns]

Comprobación de duplicados

In [36]:
df.duplicated().sum()

0

No tenemos duplicados

### Columnas numéricas

In [28]:
df.describe().T.round(2)

count      mean       std      min       25%       50%  \
RowID         51290.0  25645.50  14806.29     1.00  12823.25  25645.50   
PostalCode     9994.0  55190.38  32063.69  1040.00  23223.00  56430.50   
Sales         51290.0    246.49    487.57     0.44     30.76     85.05   
Quantity      51290.0      3.48      2.28     1.00      2.00      3.00   
Discount      51290.0      0.14      0.21     0.00      0.00      0.00   
Profit        51290.0     28.61    174.34 -6599.98      0.00      9.24   
ShippingCost  51290.0     26.38     57.30     0.00      2.61      7.79   

                   75%       max  
RowID         38467.75  51290.00  
PostalCode    90008.00  99301.00  
Sales           251.05  22638.48  
Quantity          5.00     14.00  
Discount          0.20      0.85  
Profit           36.81   8399.98  
ShippingCost     24.45    933.57

`RowID`: Tenemos un total de 51290 valores que se corresponden con un identificador único de las filas.

`PostalCode`: Código postal. Tenemos muchos valores nulos y no nos aporta información relevante ya que necesitaríamos contexto adicional.

`Sales`: Dinero en ventas. Tiene una gran dispersión ya que toma valores desde menores a 1 hasta más de 20000, con un promedio de 246.49.

`Quantity`: Valor entero que va desde 1 unidad hasta un máximo de 14, con un promedio de entre 3 y 4 unidades por operación.

`Discount`: Porcentaje de descuento en tanto por 1. Tiene un mínimo de 0, que se correspondería con entradas sin descuentos hasta un valor de 0.85, que se corresponde con un descuento del 85%. El descuento promedio de 14%.

`Profit`: Beneficio obtenido de la operación. El promedio es de 28.61 con un mínimo de -6600 (lo que indica pérdidas), un máximo de 8400, y una dispersión de 174.

`ShippingCost`: El coste de envío. Tiene un máximo de 933 y un mínimo de 0, lo que indica que hay operaciones que no tienen envío o que su coste es gratuito. Su dispersión es más reducida, de 57 unidades.

### Columnas categóricas

In [33]:
for col in df.select_dtypes(include=['O', 'category']).columns.to_list():
    print(f"Current column: {col}")
    print("-" * 50)
    display(value_counts(df, col).head())
    checker(df, col)
    print("\n\n")

Current column: OrderID
--------------------------------------------------
The number of unique values for this category is 25035


count  proportion
OrderID                          
CA-2014-100111     14         0.0
NI-2014-8880       13         0.0
TO-2014-9950       13         0.0
IN-2012-41261      13         0.0
IN-2013-42311      13         0.0

Number of entries: 51290.
Number of OrderID distinct entries: 25035.
Number of OrderID duplicated: 26255.
Number of OrderID null: 0.



Current column: OrderDate
--------------------------------------------------
The number of unique values for this category is 1430


count  proportion
OrderDate                    
18-06-2014    135         0.0
18-11-2014    127         0.0
03-09-2014    126         0.0
20-11-2014    118         0.0
29-12-2014    116         0.0

Number of entries: 51290.
Number of OrderDate distinct entries: 1430.
Number of OrderDate duplicated: 49860.
Number of OrderDate null: 0.



Current column: ShipDate
--------------------------------------------------
The number of unique values for this category is 1464


count  proportion
ShipDate                     
22-11-2014    130         0.0
07-09-2014    115         0.0
07-12-2014    101         0.0
17-11-2014    101         0.0
29-11-2014    100         0.0

Number of entries: 51290.
Number of ShipDate distinct entries: 1464.
Number of ShipDate duplicated: 49826.
Number of ShipDate null: 0.



Current column: ShipMode
--------------------------------------------------
The number of unique values for this category is 4


count  proportion
ShipMode                         
Standard Class  30775        0.60
Second Class    10309        0.20
First Class      7505        0.15
Same Day         2701        0.05

Number of entries: 51290.
Number of ShipMode distinct entries: 4.
Number of ShipMode duplicated: 51286.
Number of ShipMode null: 0.



Current column: CustomerID
--------------------------------------------------
The number of unique values for this category is 1590


count  proportion
CustomerID                   
PO-18850       97         0.0
BE-11335       94         0.0
JG-15805       90         0.0
SW-20755       89         0.0
EM-13960       85         0.0

Number of entries: 51290.
Number of CustomerID distinct entries: 1590.
Number of CustomerID duplicated: 49700.
Number of CustomerID null: 0.



Current column: CustomerName
--------------------------------------------------
The number of unique values for this category is 795


count  proportion
CustomerName                      
Muhammed Yedwab    108         0.0
Steven Ward        106         0.0
Gary Hwang         102         0.0
Patrick O'Brill    102         0.0
Bill Eplett        102         0.0

Number of entries: 51290.
Number of CustomerName distinct entries: 795.
Number of CustomerName duplicated: 50495.
Number of CustomerName null: 0.



Current column: Segment
--------------------------------------------------
The number of unique values for this category is 3


count  proportion
Segment                       
Consumer     26518        0.52
Corporate    15429        0.30
Home Office   9343        0.18

Number of entries: 51290.
Number of Segment distinct entries: 3.
Number of Segment duplicated: 51287.
Number of Segment null: 0.



Current column: City
--------------------------------------------------
The number of unique values for this category is 3636


count  proportion
City                            
New York City    915        0.02
Los Angeles      747        0.01
Philadelphia     537        0.01
San Francisco    510        0.01
Santo Domingo    443        0.01

Number of entries: 51290.
Number of City distinct entries: 3636.
Number of City duplicated: 47654.
Number of City null: 0.



Current column: State
--------------------------------------------------
The number of unique values for this category is 1094


count  proportion
State                           
California      2001        0.04
England         1499        0.03
New York        1128        0.02
Texas            985        0.02
Ile-de-France    981        0.02

Number of entries: 51290.
Number of State distinct entries: 1094.
Number of State duplicated: 50196.
Number of State null: 0.



Current column: Country
--------------------------------------------------
The number of unique values for this category is 147


count  proportion
Country                         
United States   9994        0.19
Australia       2837        0.06
France          2827        0.06
Mexico          2644        0.05
Germany         2065        0.04

Number of entries: 51290.
Number of Country distinct entries: 147.
Number of Country duplicated: 51143.
Number of Country null: 0.



Current column: Market
--------------------------------------------------
The number of unique values for this category is 7


count  proportion
Market                   
APAC    11002        0.21
LATAM   10294        0.20
EU      10000        0.19
US       9994        0.19
EMEA     5029        0.10

Number of entries: 51290.
Number of Market distinct entries: 7.
Number of Market duplicated: 51283.
Number of Market null: 0.



Current column: Region
--------------------------------------------------
The number of unique values for this category is 13


count  proportion
Region                    
Central  11117        0.22
South     6645        0.13
EMEA      5029        0.10
North     4785        0.09
Africa    4587        0.09

Number of entries: 51290.
Number of Region distinct entries: 13.
Number of Region duplicated: 51277.
Number of Region null: 0.



Current column: ProductID
--------------------------------------------------
The number of unique values for this category is 10292


count  proportion
ProductID                         
OFF-AR-10003651     35         0.0
OFF-AR-10003829     31         0.0
OFF-BI-10003708     30         0.0
OFF-BI-10002799     30         0.0
FUR-CH-10003354     28         0.0

Number of entries: 51290.
Number of ProductID distinct entries: 10292.
Number of ProductID duplicated: 40998.
Number of ProductID null: 0.



Current column: Category
--------------------------------------------------
The number of unique values for this category is 3


count  proportion
Category                          
Office Supplies  31273        0.61
Technology       10141        0.20
Furniture         9876        0.19

Number of entries: 51290.
Number of Category distinct entries: 3.
Number of Category duplicated: 51287.
Number of Category null: 0.



Current column: Sub-Category
--------------------------------------------------
The number of unique values for this category is 17


count  proportion
Sub-Category                   
Binders        6152        0.12
Storage        5059        0.10
Art            4883        0.10
Paper          3538        0.07
Chairs         3434        0.07

Number of entries: 51290.
Number of Sub-Category distinct entries: 17.
Number of Sub-Category duplicated: 51273.
Number of Sub-Category null: 0.



Current column: ProductName
--------------------------------------------------
The number of unique values for this category is 3788


count  proportion
ProductName                                      
Staples                           227         0.0
Cardinal Index Tab, Clear          92         0.0
Eldon File Cart, Single Width      90         0.0
Rogers File Cart, Single Width     84         0.0
Ibico Index Tab, Clear             83         0.0

Number of entries: 51290.
Number of ProductName distinct entries: 3788.
Number of ProductName duplicated: 47502.
Number of ProductName null: 0.



Current column: OrderPriority
--------------------------------------------------
The number of unique values for this category is 4


count  proportion
OrderPriority                   
Medium         29433        0.57
High           15501        0.30
Critical        3932        0.08
Low             2424        0.05

Number of entries: 51290.
Number of OrderPriority distinct entries: 4.
Number of OrderPriority duplicated: 51286.
Number of OrderPriority null: 0.





`OrderID`: Identificador del pedido. Vemos que tenemos varios pedidos con la misma entrada por lo que pueden contener diferentes operaciones. Tenemos un total de 25035 pedidos.

`OrderDate`: Tenemos fechas que constan de varias operaciones y, posiblemente, a varios pedidos y clientes. Hay un total de 1430 días diferentes.

`ShipDate`: Al igual que el campo anterior, tenemos fechas que constan de varias operaciones y, posiblemente, a varios pedidos y clientes. Hay un total de 1464 días diferentes.

`ShipMode`: Tenemos un total de 4 modos de envío. El 60% de los envíos se llevan a cabo mediante `StandardClass` con tan solo un 5% conrrespondiendo a `Same Day`, que podríamos considerar como máxima urgencia.

`CustomerID`: Identificador único de cada cliente. Tenemos un total de 1590 clientes diferentes.

`CustomerName`: Nombre de cada cliente. En este caso solo hay 795 nombres únicos. Esto sugiere que puede haber clientes diferentes que se llamen igual, lo cual es perfectamente posible, especialmente en nombrres comunes. Eliminaremos esta columna ya que nos quedaremos con la información de los ID.

`Segment`: Segmento al que pertenece cada operación. Hay 3 opciones: `Consumer`, `Corporate`, `Home Office`, donde `Consumer`es la más frecuente con un 52% de las entradas.Tendremos que ver si es una característica de los tipos de cliente.

`City`: Ciudad en la que tiene lugar la operación. Hay un total de 3635 diferentes.

`State`: Estado en el que tiene lugar la operación. Tenemos un total de 1094 diferentes.

`Country`: País en el que tiene lugar la operación. Tenemos un total de 147.

`Market`: Los diferentes mercados en los que tienen lugar las diferentes operaciones. Tenemos un total de 7 diferentes que engloban a los países mencionados.

`Region`: La región mundial o continental en la que se encuentra la operación. Tenemos un total de 13 deifrentes.

`ProductID`: Identificador único del cada producto. Tenemos un total de 10292 productos.

`Category`: Categoría a la que corresponde el producto. Tenemos tres opciones diferentes: `Office Supplies`, `Technology` y `Furniture`.

`Sub-category`: Subcategoría a la que pertenecen los productos. Tenemos 17 diferentes.

`ProductName`: Nombre del producto. Tenemos 3788 diferentes, por lo que es posible que pueda haber productos diferentes (diferente identificador) con el mismo nombre.

`OrderPriority`: Prioridad del pedido entre 4 opciones disponibles: `Medium`, `High`, `Critical`, `Low`.

---

Vamos a comenzar eliminando las columnas que no nos aportan información útil.

`

---

Analicemos las fechas con más detalle

Análisis de `Segment`

Análisis de `Product` (categoría y sub-categoría)